### PR TITLE
feat(submit_transaction) : refactor submit transaction command

### LINKED ISSUE #55: [REFACTOR] Refine the NewSubmitTransactionCommandHandler implementation
## Desired Solution
The current implementation is missing the `x-topics` header parsing logic and the construction of tagged BEEF. Please provide the missing parts of the functionality, along with unit tests to verify all positive and negative paths of the request  processing flow.

## Suggested Implementation

Code ref link - https://github.com/bitcoin-sv/overlay-express/blob/master/src/OverlayExpress.ts#L673

# Testing 

Please ensure that all positive and negative scenarios are covered by unit tests.

### PR DESCRIPTION
**Pull request by cain4chain**
## Description of Changes

chore(submit_transaction) : add error msg and x-topics header
chore(submit_transaction_handler): add data and type validaiton

## Linked Issues / Tickets

Closes : [REFACTOR] Refine the NewSubmitTransactionCommandHandler implementation #55

## Testing Procedure

chore(submit_transaction_handler_test) : add cases for success, http type, invalid, missing and nil

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have run the linter

### CODE REVIEW

File: pkg/server/app/commands/submit_transaction.go [hunk 1]
```diff
@@ -2,13 +2,27 @@ package commands
...
+	// Error messages
+	errMsgMissingTopicsHeader = "Missing x-topics header"
+	errMsgInvalidTopicsFormat = "Invalid x-topics header format"
+	errMsgFailedToReadBody    = "Failed to read request body"
+	errMsgMethodNotAllowed    = "Method not allowed"
+
+	// Header keys
+	headerTopics = "x-topics"
+)
+
```
Comment (mgosek-4chain) [inline pkg/server/app/commands/submit_transaction.go]:
```
const XTopicsHeader = "x-topics"

const RequestBodyLimit1GB = 1000 * 1024 * 1024

var (
	ErrMissingXTopicsHeader       = errors.New("missing x-topics header")
	ErrInvalidXTopicsHeaderFormat = errors.New("invalid x-topics header format")
	ErrInvalidHTTPMethod          = errors.New("invalid HTTP method")
	ErrRequestBodyRead            = errors.New("failed to read request body")
	ErrRequestBodyTooLarge        = errors.New("request body too large")
)
```

File: pkg/server/app/commands/submit_transaction.go [hunk 2]
```diff
@@ -35,14 +47,76 @@ type SubmitTransactionHandler struct {
...
+	// 8. If the callback hasn't been triggered yet, send the response immediately
+	responseSync.Lock()
+	defer responseSync.Unlock()
+
+	if !responseSent {
+		responseSent = true
+		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: steak})
+	}
 }
```
Comment (mgosek-4chain) [inline pkg/server/app/commands/submit_transaction.go]: First and foremost, the current code has a resource leak (also present in other handlers - my fault 😅) due to not closing the response body after processing. Accumulated unclosed response bodies can degrade performance, cause memory leaks by holding unused file descriptors, and introduce other defects. 

To fix this, we need to add `defer res.Body.Close()` as the first line in the handler. 

Another issue is reading the entire request payload into a `byte` slice without setting a request body limit. This can be easily fixed using `io.LimitReader`. However, the downside of this approach is that it loads the entire request body into memory. If multiple users send requests to our server simultaneously, we could quickly run out of RAM 😿 . That's why we should process the request body in chunks. 

The original `overlay-express` sets this limit to `1GB` - let's use that as a reference. 

```
type SubmitTransactionHandler struct {
	provider         SubmitTransactionProvider
	requestBodyLimit int64
}
```

As an additional improvement, I strongly recommend extracting `overlay.TaggedBEEF` creation steps into a separate function to encapsulate everything in one place. 

```
func (s *SubmitTransactionHandler) CreateTaggedBEEFFromRequest(r *http.Request) (*overlay.TaggedBEEF, error) {
	header := r.Header.Get(XTopicsHeader)
	if header == "" {
		return nil, ErrMissingXTopicsHeader
	}

	var topics []string
	if err := json.Unmarshal([]byte(header), &topics); err != nil {
		return nil, ErrInvalidXTopicsHeaderFormat
	}

	reader := io.LimitReader(r.Body, s.requestBodyLimit)
	buff := make([]byte, 64*1024) // typically it's a best balance between performance and memory usage.
	bytesRead := 0
	for {
		n, err := reader.Read(buff)
		bytesRead += n

		if bytesRead > int(s.requestBodyLimit) {
			return nil, ErrRequestBodyTooLarge
		}

		if err == io.EOF {
			break
		}

		if err != nil {
			return nil, ErrRequestBodyRead
		}
	}

	return &overlay.TaggedBEEF{Beef: buff, Topics: topics}, nil
}
```

The handler code looks much simpler now. The last part is ensuring that the callback function is called when the `steak` is ready. The issue lies in the design of the `overlay-engine`, where we don't know whether the callback function will be called asynchronous or synchronous. 

A weak point of `http.ResponseWriter` is that it's not thread-safe, which can lead to unpredictable behavior. Instead, we should use a channel, as it ensures only a single coroutine writes to `w`. Additionally, to avoid deadlocks or hanging requests, we can use a timeout with a custom value. 

```
func (s *SubmitTransactionHandler) Handle(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	if r.Method != http.MethodPost {
		http.Error(w, ErrInvalidHTTPMethod.Error(), http.StatusMethodNotAllowed)
		return
	}

	taggedBEEF, err := s.CreateTaggedBEEFFromRequest(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	steakChan := make(chan *overlay.Steak, 1)
	_, err = s.provider.Submit(r.Context(), *taggedBEEF, engine.SubmitModeCurrent, func(steak *overlay.Steak) {
		steakChan <- steak
	})
	if err != nil {
		jsonutil.SendHTTPResponse(w, http.StatusInternalServerError, err.Error())
		return
	}

	select {
	case steak := <-steakChan:
		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: *steak})
	case <-time.After(5 * time.Second): // This select case can be refined later, for testing purposes, let's leave it as it is.
		http.Error(w, http.StatusText(http.StatusRequestTimeout), http.StatusRequestTimeout)
	}
}
```

```
func NewSubmitTransactionCommandHandler(provider SubmitTransactionProvider) *SubmitTransactionHandler {
	if provider == nil {
		panic("submit transaction provider is nil")
	}
	return &SubmitTransactionHandler{
		provider:         provider,
		requestBodyLimit: RequestBodyLimit1GB,
	}
}
```


After all, please apply the suggestions described above. Additionally, adjust the unit tests and add any missing ones. For example, verify if the API behaves correctly when attempting to exceed the request payload.

Comment (cain4chain) [reply pkg/server/app/commands/submit_transaction.go]: Thanks for all the details, I will update this as requested.

File: pkg/server/app/commands/submit_transaction.go [hunk 3]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
 // sends a JSON response after invoking the engine and returns an HTTP response
 // with the appropriate status code based on the engine's response.
 func (s *SubmitTransactionHandler) Handle(w http.ResponseWriter, r *http.Request) {
-	// TODO: Add custom validation logic.
-	steak, err := s.provider.Submit(r.Context(), overlay.TaggedBEEF{}, engine.SubmitModeCurrent, func(steak *overlay.Steak) {})
+	defer func() {
+		if err := r.Body.Close(); err != nil {
+			fmt.Printf("Failed to close request body: %v\n", err)
+		}
+	}()
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go]:
```suggestion
	defer defer func() {
			_ = r.Body.Close()
		}()
```
this will pass linter

File: pkg/server/app/commands/submit_transaction_test.go [hunk 4]
```diff
@@ -1,45 +1,253 @@
...
+func (SubmitTransactionProviderNeverCallback) Submit(ctx context.Context, taggedBEEF overlay.TaggedBEEF, mode engine.SumbitMode, onSteakReady engine.OnSteakReady) (overlay.Steak, error) {
+	// Never call the callback which then should trigger the timeout
+	return overlay.Steak{}, nil
+}
+
+// For testing purposes only - allows creating a handler with a custom body limit using middleware type approach
+func createTestHandlerWithLimit(provider commands.SubmitTransactionProvider, limit int64) http.HandlerFunc {
+	handler, err := commands.NewSubmitTransactionCommandHandler(provider)
+	if err != nil {
+		panic(err)
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction_test.go]:
```suggestion
		panic(err)
```
I think now after refactoring commans and queries ... we avoiding of `panic` method .. maybe change function signature and return `(http.HandlerFunc, error)`

File: pkg/server/app/commands/submit_transaction.go [hunk 5]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
+	if header == "" {
+		return nil, ErrMissingXTopicsHeader
+	}
+
+	var topics []string
+	if err := json.Unmarshal([]byte(header), &topics); err != nil {
+		return nil, ErrInvalidXTopicsHeaderFormat
+	}
+
+	reader := io.LimitReader(r.Body, s.requestBodyLimit)
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go]:
```suggestion
	reader := io.LimitReader(r.Body, s.requestBodyLimit)
	defer func(){
	    _ =  r.Body.Close()
	}
```

File: pkg/server/app/commands/submit_transaction.go [hunk 6]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
+	buff := make([]byte, 64*1024) // typically it's a best balance between performance and memory usage (loop and size).
+	bytesRead := 0
+	for {
+		n, err := reader.Read(buff)
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go:91]: i think  the `buff` slice is overwritten by each call of `reader.Read(buff)` 
```go
n, err := reader.Read(buff)
if n > 0 {
    if _, writeErr := beefBuffer.Write(buff[:n]); writeErr != nil {
        return nil, ErrRequestBodyRead
    }
}
``` 
this would save every chunk of the body into a buffer `beefBuffer`

File: pkg/server/app/commands/submit_transaction.go [hunk 7]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
+// CreateTaggedBEEFFromRequest extracts the topics from the header and reads the body
+// to create a TaggedBEEF object, with size limits applied.
+func (s *SubmitTransactionHandler) CreateTaggedBEEFFromRequest(r *http.Request) (*overlay.TaggedBEEF, error) {
+	header := r.Header.Get(XTopicsHeader)
+	if header == "" {
+		return nil, ErrMissingXTopicsHeader
+	}
+
+	var topics []string
+	if err := json.Unmarshal([]byte(header), &topics); err != nil {
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go]: the header is slice ? `X-Topics: ["topic1","topic2","topic3"]` would not be better to have `X-Topics: topic1,topic2,topic3` and split this via comma ? 
```go
header := r.Header.Get(XTopicsHeader)
if header == "" {
    return nil, ErrMissingXTopicsHeader
}

topics := strings.Split(header, ",")
```

if we have control of how header is configured :)

File: pkg/server/app/commands/submit_transaction.go [hunk 8]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
+
+	if err != nil {
+		jsonutil.SendHTTPResponse(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	select {
+	case steak := <-steakChan:
+		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: *steak})
+	case <-time.After(5 * time.Second): // Timeout to prevent hanging requests
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go]: 5 seconds is okay for now, but you might want to make it configurable (like s.steakResponseTimeout) ?

File: pkg/server/app/commands/submit_transaction.go [hunk 9]
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
...
+		http.Error(w, http.StatusText(http.StatusRequestTimeout), http.StatusRequestTimeout)
```
Comment (ac4ch) [inline pkg/server/app/commands/submit_transaction.go:132]: i think if timeout occurs, it could be good to log the timeout event for monitoring/debugging.

File: pkg/server/app/commands/submit_transaction.go (hunk 3 above)
Comment (cain4chain) [reply pkg/server/app/commands/submit_transaction.go]: ok thanks 👍

File: pkg/server/app/commands/submit_transaction.go (hunk 6 above)
Comment (cain4chain) [reply pkg/server/app/commands/submit_transaction.go:91]: yes you are right, thanks I will update this

File: pkg/server/app/commands/submit_transaction.go (hunk 8 above)
Comment (cain4chain) [reply pkg/server/app/commands/submit_transaction.go]: Ok will create a public function for this and make it 5 seconds for default

**Review by mgosek-4chain — COMMENTED** [review]
I've added suggestions how we can improve the current code base, please take a look :) 

https://github.com/4chain-ag/go-overlay-services/tree/REFACTOR-Refine-the-NewSubmitTransactionCommandHandler-cr-suggestions

File: pkg/server/app/commands/submit_transaction.go [hunk 10]
```diff
@@ -29,30 +55,134 @@ type SubmitTransactionProvider interface {
...
 
 // Handle orchestrates the processing flow of a transaction. It prepares and
 // sends a JSON response after invoking the engine and returns an HTTP response
 // with the appropriate status code based on the engine's response.
 func (s *SubmitTransactionHandler) Handle(w http.ResponseWriter, r *http.Request) {
-	// TODO: Add custom validation logic.
-	steak, err := s.provider.Submit(r.Context(), overlay.TaggedBEEF{}, engine.SubmitModeCurrent, func(steak *overlay.Steak) {})
+	defer func() {
+		_ = r.Body.Close()
+	}()
```
Comment (mgosek-4chain) [inline pkg/server/app/commands/submit_transaction.go]: This can be removed because we're already using a `SafeHandler` that handles closing the request bodies :)

**Review by mgosek-4chain — APPROVED** [review]

File: pkg/server/app/commands/submit_transaction.go (hunk 10 above)
Comment (cain4chain) [reply pkg/server/app/commands/submit_transaction.go]: Thanks, I will update this.
